src/mame/heathkit/tlb.cpp

```diff
@@ -0,0 +1,751 @@
+/***************************************************************************
+ Memory Layout
+	 The U435 three-to-eight line decoder uses A14 and A15 to generate three memory addresses:
+
+	 1.   Program ROM        0x0000
+
+	 2.   Scratchpad RAM     0x4000
+
+	 3.   Display Memory     0xF800
+
+
+ Port Layout
+
+	 Only address lines A5, A6, A7 are used by the U442 three-to-eight line decoder
+
+Address   Description
+----------------------------------------------------
+ 0x00   Power-up configuration (primary - SW401)
+ 0x20   Power-up configuration (secondary - SW402)
+ 0x40   ACE (communications)
+ 0x60   CRT controller
+ 0x80   Keyboard encoder
+ 0xA0   Keyboard status
+ 0xC0   Key click enable
+ 0xE0   Bell enable
+
+****************************************************************************/
+
+#include "emu.h"
+
+#include "tlb.h"
+
+// Standard H19 used a 2.048 MHz clock
+#define H19_CLOCK (XTAL(12'288'000) / 6)
+#define MC6845_CLOCK (XTAL(12'288'000) /8)
+#define INS8250_CLOCK (XTAL(12'288'000) /4)
+
+// Capacitor value in pF
+#define H19_KEY_DEBOUNCE_CAPACITOR 5000
+#define MM5740_CLOCK (mm5740_device::calc_effective_clock_key_debounce(H19_KEY_DEBOUNCE_CAPACITOR))
+
+// Beep Frequency is 1 KHz
+#define H19_BEEP_FRQ (H19_CLOCK / 2048)
+
+DEFINE_DEVICE_TYPE(HEATH_TLB, heath_tlb_device, "heath_tlb", "Heath Terminal Logic Board");
+DEFINE_DEVICE_TYPE(HEATH_SUPER19, heath_super19_tlb_device, "heath_super19_tlb", "Heath Terminal Logic Board w/Super19 ROM");
+DEFINE_DEVICE_TYPE(HEATH_WATZ, heath_watz_tlb_device, "heath_watz_tlb", "Heath Terminal Logic Board w/Watzman ROM");
+DEFINE_DEVICE_TYPE(HEATH_ULTRA, heath_ultra_tlb_device, "heath_ultra_tlb", "Heath Terminal Logic Board w/Ultra ROM");
+
+heath_tlb_device::heath_tlb_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	heath_tlb_device(mconfig, HEATH_TLB, tag, owner, clock)
+{
+}
+
+heath_tlb_device::heath_tlb_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, type, tag, owner, clock),
+	m_write_sd(*this),
+	m_palette(*this, "palette"),
+	m_maincpu(*this, "maincpu"),
+	m_crtc(*this, "crtc"),
+	m_ace(*this, "ins8250"),
+	m_beep(*this, "beeper"),
+	m_p_videoram(*this, "videoram"),
+	m_p_chargen(*this, "chargen"),
+	m_mm5740(*this, "mm5740"),
+	m_kbdrom(*this, "keyboard"),
+	m_kbspecial(*this, "MODIFIERS")
+{
+}
+
+void heath_tlb_device::checkBeepState()
+{
+	if (!m_keyclickactive && !m_bellactive)
+	{
+		m_beep->set_state(0);
+	}
+}
+
+TIMER_CALLBACK_MEMBER(heath_tlb_device::key_click_off)
+{
+	m_keyclickactive = false;
+
+	checkBeepState();
+}
+
+TIMER_CALLBACK_MEMBER(heath_tlb_device::bell_off)
+{
+	m_bellactive = false;
+
+	checkBeepState();
+}
+
+void heath_tlb_device::mem_map(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0000, 0x1fff).mirror(0x2000).rom();
+	map(0x4000, 0x4100).mirror(0x3e00).ram();
+	map(0xc000, 0xc7ff).mirror(0x3800).ram().share("videoram");
```

Comment:
You should use your object finder here rather than the literal tag – it helps avoid having it get out-of-sync.